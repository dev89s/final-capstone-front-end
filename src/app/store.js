import { configureStore } from '@reduxjs/toolkit';

import dataReducer from '../components/Home/createSlice';
import userReducer from './user/userSlice';
import roomReducer from './room/roomSlice';

const store = configureStore({
  reducer: {
    user: userReducer,
    dataSlice: dataReducer,
    room: roomReducer,
  },
});

export default store;
