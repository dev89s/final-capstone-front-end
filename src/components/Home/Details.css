.details-container {
  display: flex;
  flex-direction: row;
  height: 74vh;
  width: 83vw;
  justify-content: space-between;
  padding: 3vw 3vh;
}

img {
  width: 48vw;
  height: 65vh;
}

.nam-room {
  display: flex;
  flex-direction: column;
  align-items: end;
  width: 25vw;
  margin-bottom: 2vh;
}

table.details-table {
  width: 25vw;
}

.details-table tr:nth-child(odd) {
  background-color: #f0f0f0; /* Grey background for odd rows */
}

.details-table td {
  padding: 0.5rem;
  text-align: right;
}

.describe-details {
  display: flex;
  flex-direction: column;
  margin-top: 3vh;
  align-items: flex-end;
  gap: 8vh;
}

.describe-details a {
  text-decoration: none;
  color: white;
  height: 7vh;
  width: 9vw;
  display: flex;
  justify-content: center;
  align-items: center;
  background-color: green;
  border: none;
  border-radius: 20px;
}

.back-b a {
  text-decoration: none;
  color: white;
  height: 8vh;
  width: 6vw;
  display: flex;
  justify-content: center;
  align-items: center;
  background-color: green;
  border: none;
  border-top-right-radius: 20px;
  border-bottom-right-radius: 20px;
}
