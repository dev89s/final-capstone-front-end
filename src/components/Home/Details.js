import React from 'react';
import { useSelector } from 'react-redux';
import { useParams } from 'react-router';
import { Link } from 'react-router-dom';

const Detail = () => {
  const { roomId } = useParams();
  const p = parseInt(roomId, 10);

  const data = useSelector((state) => state.dataSlice.data);

  const room = data.find((room) => room.id === p);
  // a

  if (!room) {
    return <div>Room not found.</div>;
  }

  return (
    <div>
      <h1>
        Room Details for Room
        {' '}
        {roomId}
      </h1>
      <div>
        <img src={room.image} alt={room.name} />
        <h2>{room.name}</h2>
        <p>{room.description}</p>
        <p>{room.price}</p>
        <Link to={`/reserve/${room.id}`}>Reserve</Link>
      </div>
    </div>
  );
};

// No need for the unused prop type definition
export default Detail;
