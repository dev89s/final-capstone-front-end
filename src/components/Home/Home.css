.homepage {
  overflow-x: hidden;
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  margin: 0 3%;
  margin-top: 2%;
}

.room-container {
  display: flex;
  width: 60vw; /* Make sure the container takes full width */
  flex-wrap: nowrap; /* Prevent items from wrapping */
  overflow: hidden; /* Hide the overflowing rooms */
  transition: transform 0.5s ease; /* For smooth sliding effect */
}

.select {
  display: flex;
  flex-direction: column;
  margin-top: 14%;
  align-items: center;
}

.room-control2 button {
  height: 50px;
  margin-top: 20vh;
  font-weight: bold;
  border-top-left-radius: 20px;
  border-bottom-left-radius: 20px;
  background: greenyellow;
  border-color: greenyellow;
  color: white !important;
}

.room-control1 button {
  height: 50px;
  border-top-right-radius: 20px;
  border-bottom-right-radius: 20px;
  margin-top: 20vh;
  font-weight: bold;
  background: greenyellow;
  border-color: greenyellow;
  color: white !important;
}

.room-item {
  display: flex;
  flex-direction: column;
  flex: 0 0 auto;
  align-items: center;
  width: 300px; /* Set the width of each room item */
  margin-right: 16px;
  transition: transform 0.5s ease; /* For smooth sliding effect */

  /* margin: 20vh; */
}

.room-item img {
  width: 200px;
  height: 200px;
  border-radius: 100%;
}

@media (max-width: 600px) {
  .room-item {
    margin-right: 0; /* Remove margin for single room on smaller screens */
  }

  .room-control1,
  .room-control2 {
    display: none;
  }

  .room-container {
    flex-direction: column;
    width: 100vw;
    align-items: center;
  }
}
