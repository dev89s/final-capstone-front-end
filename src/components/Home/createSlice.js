import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import axios from 'axios';

// Create an asynchronous thunk to fetch rooms data
export const fetchRooms = createAsyncThunk('data/fetchRooms', async () => {
  const response = await axios.get('http://localhost:3000/api/v1/rooms');
  return response.data;
});

const dataSlice = createSlice({
  name: 'data',
  initialState: {
    data: [],
    status: 'idle',
    error: null,
  },
  reducers: {},
  extraReducers: (builder) => {
    builder
      .addCase(fetchRooms.pending, (state) => {
        state.status = 'loading';
      })
      .addCase(fetchRooms.fulfilled, (state, action) => {
        state.status = 'succeeded';
        state.data = action.payload;
      })
      .addCase(fetchRooms.rejected, (state, action) => {
        state.status = 'failed';
        state.error = action.error.message;
      });
  },
});

export default dataSlice.reducer;
