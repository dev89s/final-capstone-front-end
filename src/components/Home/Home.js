import React, { useEffect, useRef } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import './Home.css';
import { Link } from 'react-router-dom';
import { fetchRooms } from './createSlice';

const Home = () => {
  const dispatch = useDispatch();
  const data = useSelector((state) => state.dataSlice.data);

  useEffect(() => {
    dispatch(fetchRooms());
  }, [dispatch]);

  const scrollContainerRef = useRef(null);

  const slideRight = () => {
    if (scrollContainerRef.current) {
      const roomWidth = scrollContainerRef.current.querySelector('.room-item').offsetWidth;
      const currentTransform = scrollContainerRef.current.style.transform;
      const currentTranslateX = parseInt(currentTransform.replace('translateX(', '').replace('px)', ''), 10);
      const newTranslateX = currentTranslateX - roomWidth * 3;
      scrollContainerRef.current.style.transform = `translateX(${newTranslateX}px)`;
    }
  };

  const slideLeft = () => {
    if (scrollContainerRef.current) {
      scrollContainerRef.current.style.transform = 'translateX(0)';
    }
  };

  return (
    <>
      <div className="select">
        <h1>LATEST ROOMS</h1>
        <p>Please select a room</p>
      </div>
      <div className="homepage">
        <div className="room-control1">
          <button type="button" onClick={slideLeft}>Slide Left</button>
        </div>
        <div className="room-container" ref={scrollContainerRef}>
          {data.map((room) => (
            <Link to={`/detail/${room.id}`} key={room.id} className="room-item">
              <img src={room.image} alt={room.name} />
              <h1>{room.name}</h1>
              <p>{room.description}</p>
            </Link>
          ))}
        </div>
        <div className="room-control2">
          <button type="button" onClick={slideRight}>Slide Right</button>
        </div>
      </div>
    </>
  );
};

export default Home;
