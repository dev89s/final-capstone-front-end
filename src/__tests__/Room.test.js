import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import Home from '../components/Home/Home';
import store from '../app/store';

const initialState = {
  roomList: [],
  isLoading: false,
  error: undefined,
};

const data = {
  rooms: [
    {
      room_id: '1',
      room_name: 'Room 1',
      description: 'The Room 1 description',
      price: 234,
      image: 'images',
    },
    {
      room_id: '2',
      room_name: 'Room 2',
      description: 'The Room 2 description',
      price: 2234,
      image: 'images',
    },
  ],
};

it('should render the Home page', () => {
  const tree = renderer
    .create(
      <Provider store={store}>
        <Home />
      </Provider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it('should render the Room', () => {
  const mockStore = configureStore();
  const RoomStore = mockStore(initialState);

  RoomStore.clearActions();
  const getRooms = () => ({ type: 'GET_ROOMS' });

  RoomStore.dispatch(getRooms());

  const actions = RoomStore.getActions();
  const expectedPayload = { type: 'GET_ROOMS' };
  expect(actions).toEqual([expectedPayload]);
});

it('should render the initial state Properly', () => {
  const mockStore = configureStore();
  const RoomStore = mockStore(initialState);

  expect(initialState).toEqual(RoomStore.getState());
});

it('should render the Room state Properly', () => {
  const mockStore = configureStore();

  const RoomStore = mockStore(data);

  expect(data).toEqual(RoomStore.getState());
});
