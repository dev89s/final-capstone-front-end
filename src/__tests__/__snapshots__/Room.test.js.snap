// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render the Home page 1`] = `
Array [
  <div
    className="select"
  >
    <h1>
      LATEST ROOMS
    </h1>
    <p>
      Please select a room
    </p>
  </div>,
  <div
    className="homepage"
  >
    <div
      className="room-control1"
    >
      <button
        className="disabledButton"
        disabled={true}
        onClick={[Function]}
        type="button"
      >
        &lt;
      </button>
    </div>
    <div
      className="room-container"
    />
    <div
      className="room-control2"
    >
      <button
        className="false"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        &gt;
      </button>
    </div>
  </div>,
]
`;
